### PR TITLE
Add nonnull annotation

### PR DESCRIPTION
**Pull request by casz**
Let's have some annotation for these methods
Upstream of https://github.com/jenkinsci/blueocean-display-url-plugin/pull/13

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/displayurlapi/ClassicDisplayURLProvider.java [hunk 1]
```diff
@@ -4,6 +4,7 @@
 import hudson.Util;
 import hudson.model.Job;
 import hudson.model.Run;
+import javax.annotation.Nonnull;
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/ClassicDisplayURLProvider.java]: Where is the JSR release defining this `javax` namespaced annotation? It is very dangerous to use annotations in `javax` unless they have been *published* through the JSR process. I would advise against using this annotation unless you can point to an official JSR release, especially given that the java module system does not allow two modules to share the same package names and IIRC an existing module is already publishing the `javax.annotation` package name (the module is `java.xml.ws.annotation`) so using this annotation would also cause issues with migrating Jenkins to support the module path.

In short, unless and until there is a JSR that *publishes* this annotation it is extremely limiting on the evolution of Jenkins to continue using it. Use the one from spotbugs instead: https://static.javadoc.io/com.github.spotbugs/spotbugs-annotations/3.1.8/edu/umd/cs/findbugs/annotations/NonNull.html

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/ClassicDisplayURLProvider.java]: it was used previously in the package so I thought, what the heck but I totally agree 🤭

**Comment by darxriggs** [comment]
@casz there are conflicts to resolve now.

**Comment by casz** [comment]
Sweet, I will resolve since Java 8/11 support is finally here 😄

**Comment by casz** [comment]
@dwnusbaum and @darxriggs please take a look 😄

File: src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java [hunk 2]
```diff
@@ -88,7 +88,7 @@ private DisplayURLContext(DisplayURLContext parent) {
      */
     private void guessPlugin() {
         StackTraceElement[] stack = (new Throwable()).getStackTrace();
-        PluginManager manager = Jenkins.getActiveInstance().getPluginManager();
+        PluginManager manager = Jenkins.getInstance().getPluginManager();
```
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java]: Is `Jenkins.get()` already available in the used Jenkins version? Then this should be used.

File: src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java [hunk 3]
```diff
@@ -45,27 +44,21 @@ public final Object doRedirect(StaplerRequest req, StaplerResponse rsp) throws I
...
-
-                    return displayURLProvider.getName().equals(providerName);
-                }
-            });
-
-            DisplayURLProvider provider = Iterables.getFirst(filtered, null);
+        if(StringUtils.isNotEmpty(providerName)) {
+            ExtensionList<DisplayURLProvider> providers = DisplayURLProvider.all();
+            DisplayURLProvider provider = providers.stream()
+                .filter(displayURLProvider -> providerName.equals(displayURLProvider.getName()))
```
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java]: Handling `displayURLProvider == null` is not required anymore?

File: src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java (hunk 2 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLContext.java]: Sadly it is not 😢

File: src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java (hunk 3 above)
Comment (casz) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java]: Nicely spotted 😄

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java]: Fixed 👍

File: src/main/java/org/jenkinsci/plugins/displayurlapi/user/PreferredProviderUserProperty.java [hunk 4]
```diff
@@ -33,22 +31,18 @@ public ProviderOption getProvider() {
...
-            public ProviderOption apply(DisplayURLProvider input) {
-                return new ProviderOption(input.getClass().getName(), input.getDisplayName());
-            }
-        });
-        return ImmutableList.copyOf(Iterables.concat(Lists.newArrayList(ProviderOption.DEFAULT_OPTION), options));
+        Stream<ProviderOption> defaultOption = Stream.of(ProviderOption.DEFAULT_OPTION);
+        Stream<ProviderOption> options = DisplayURLProvider.all().stream()
+            .map(input -> new ProviderOption(input.getClass().getName(), input.getDisplayName()));
+        Stream<ProviderOption> combined = Stream.concat(defaultOption, options);
+        return ImmutableList.copyOf(combined.collect(Collectors.toList()));
```
Comment (darxriggs) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/user/PreferredProviderUserProperty.java]: Sometimes, as in this case, I find Java's stream API too verbose and harder to read than other solutions.
This could also be written as (untested):
```
List<ProviderOption> options = DisplayURLProvider.all().stream()
        .map(input -> new ProviderOption(input.getClass().getName(), input.getDisplayName()))
        .collect(Collectors.toList());
return ImmutableList.builder().add(ProviderOption.DEFAULT_OPTION).addAll(options).build();
```

Comment (casz) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/user/PreferredProviderUserProperty.java]: indeed that reads better :smile: I just don't use Guava that much.

**Comment by casz** [comment]
@stephenc can you review so I can get your request for change, changed? 🤭
@dwnusbaum @darxriggs chance for another review?

File: src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java [hunk 5]
```diff
@@ -74,9 +69,11 @@ DisplayURLProvider lookupProvider() {
         }
         DisplayURLProvider displayURLProvider = DisplayURLProvider.getPreferredProvider();
         if (displayURLProvider == null) {
-            Iterable<DisplayURLProvider> all = DisplayURLProvider.all();
-            Iterable<DisplayURLProvider> availableProviders = Iterables.filter(all, Predicates.not(Predicates.instanceOf(ClassicDisplayURLProvider.class)));
-            displayURLProvider = Iterables.getFirst(availableProviders, DisplayURLProvider.getDefault());
+            ExtensionList<DisplayURLProvider> all = DisplayURLProvider.all();
+            displayURLProvider = all.stream()
+                .filter(((Predicate<DisplayURLProvider>) ClassicDisplayURLProvider.class::isInstance)
```
Comment (casz) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/actions/AbstractDisplayAction.java]: Java 11 will improve this with [`not(ClassicDisplayURLProvider.class::isInstance)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/function/Predicate.html#not(java.util.function.Predicate))

File: src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLProvider.java [hunk 6]
```diff
@@ -35,23 +34,25 @@ public static DisplayURLProvider get() {
      *
      * @return all the {@link DisplayURLProvider} implementations.
      */
-    public static Iterable<DisplayURLProvider> all() {
+    public static ExtensionList<DisplayURLProvider> all() {
```
Comment (casz) [inline src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLProvider.java]: This is not breaking since ExtensionList is still Iterable, it just lends itself better to stream API.

Comment (dwnusbaum) [reply src/main/java/org/jenkinsci/plugins/displayurlapi/DisplayURLProvider.java]: I'm not sure about this, it causes problems in environments like the PCT: 

```
java.lang.NoSuchMethodError: org.jenkinsci.plugins.displayurlapi.DisplayURLProvider.all()Ljava/lang/Iterable;
	at org.jenkinsci.plugins.blueoceandisplayurl.BlueOceanDisplayURLImplTest.setUp(BlueOceanDisplayURLImplTest.java:178)
```

See [BlueOceanDisplayURLImplTest](https://github.com/jenkinsci/blueocean-display-url-plugin/blob/59bcc13a2c8dff20c6866caa8b2020b039026d57/src/test/java/org/jenkinsci/plugins/blueoceandisplayurl/BlueOceanDisplayURLImplTest.java#L178). That is the only use of `DisplayURLProvider.all()` outside of this plugin that I found from a quick search, and it's in test code, so it doesn't seem to be a big deal. 

Based on the [spec](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.4.15), I don't think this is a compatible change, since JARs compiled against the old class are looking for a method whose return type is `Iterable`, but not such method exists. Because the new return type is a subtype of the old type, I think the change is _source_ compatible in that all callers can be recompiled against the new version of this plugin without needing to make any changes on their end and things will work correctly, but not _binary_ compatible.
